internal/models/sessions.go

```diff
@@ -104,6 +106,54 @@ func (s *Session) UpdateOnlyRefreshInfo(tx *storage.Connection) error {
 	return tx.UpdateOnly(s, "refreshed_at", "user_agent", "ip")
 }
 
+type SessionValidityReason = int
+
+const (
+	SessionValid        SessionValidityReason = iota
+	SessionPastNotAfter                       = iota
+	SessionPastTimebox                        = iota
+	SessionTimedOut                           = iota
+)
+
+func (s *Session) CheckValidity(now time.Time, refreshTokenTime *time.Time, timebox, inactivityTimeout *time.Duration) SessionValidityReason {
+	if s.NotAfter != nil && now.After(*s.NotAfter) {
+		return SessionPastNotAfter
+	}
+
+	if timebox != nil && *timebox != 0 && now.After(s.CreatedAt.Add(*timebox)) {
+		return SessionPastTimebox
+	}
+
+	if inactivityTimeout != nil && *inactivityTimeout != 0 && now.After(s.LastRefreshedAt(refreshTokenTime).Add(*inactivityTimeout)) {
```

Comment:
likewise for `inactivityTimeout`, if we check that it can't be 0 in the config, we don't have to check it again here

Reply:
Again, config is one thing, this function is another.